jax/_src/numpy/lax_numpy.py

```diff
@@ -2893,7 +2893,14 @@ def array(object, dtype=None, copy=True, order="K", ndmin=0):
   weak_type = dtype is None and dtypes.is_weakly_typed(object)
   dtype = dtype and dtypes.canonicalize_dtype(dtype)
 
-  if _can_call_numpy_array(object):
+  leaves = tree_leaves(object)
+  can_call_numpy_array = not _any(
+      isinstance(leaf, (core.Tracer, DeviceArray)) for leaf in leaves)
+
+  if can_call_numpy_array:
+    # Construct array with numpy, but use JAX promotion rules to determine the output dtype.
+    if dtype is None and leaves:
+      dtype = dtypes.result_type(*map(dtypes.coerce_to_array, leaves))
```

Comment:
My main concern here is that the reason for this path is so `jnp.array(...large Python value...)` is fast. We now do per-object Python work.

What does this PR do to the timing of, say, passing a large list to `jnp.array`?

If the effect is large we may ultimately need to implement our own version of this code in C++.

Reply:
I haven't benchmarked it, but `_can_call_numpy_array` already does a `tree_map` across all inputs, so I don't think an additional `tree_map` will be significant.

Reply:
My concern isn't the `tree_map`, but instead the per-element call to `coerce_to_array`.

Reply:
I see - `coerce_to_array` is much more costly than a type check, so yes this does come at a computational cost.

I suppose our tradeoff here, short of pushing array creation to C++, is speed vs. correctness.

Reply:
And knowing what the benchmark impact is on a simple microbenchmark would probably go a long way to answering how much we care.

Reply:
On master:
```python
In [3]: L = list(range(1000))
In [4]: %timeit np.array(L)
55.8 µs ± 809 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
In [5]: %timeit jnp.array(L)
345 µs ± 4.89 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
```
on this branch:
```python
In [6]: %timeit jnp.array(L)                                                                                                                                                             
5.8 ms ± 36.5 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```
So about 10x slower to call `coerce_to_array` on each element.

Reply:
What we're trying to protect against here is cases where numpy is applying its casting rules to list inputs, e.g.
```python
>>> np.array([0, 1 << 63]).dtype                                                                                                                                                     
dtype('float64')
```
Under the proposal here, this should be an overflow error. I don't know how to do that without adding some per-item Python logic... perhaps we could try for something less heavy-weight than `coerce_to_array`. Any thoughts?

Reply:
We can reduce the time by aboud a factor of 3 if we only call `result_type` rather than `_coerce_to_array`, at which point numpy's handling of large integers will lead to an OverflowError if the result is int64. But if we want overflow logic to eventually be implemented in X32 mode, that solution will no longer work.

Reply:
The other options I see are:
- implement list-to-array parsing logic into C++, which sounds pretty heavy
- accept and document the fact that `jnp.array()` called on list inputs will not have consistent behavior.

Reply:
For now I restored the previous behavior and left a TODO.